### DISCUSSION TITLE
extension-cli publish returns 403 - what are required permisions for a user to modify Asset/Project context

### DISCUSSION THREAD

**Discussion by griemens** [iTwin/itwinjs-core#1613] [Q&A]
Created a sample extension which is working with iTwin viewer example app.
 
Trying now to  use extension-cli to publish the extension to a project or asset, but fails for my user account.
Command used: 
`node extension-cli publish --contextId <project_id> --extensionName "sample" --extensionVersion "0.1.0" --path C:/14.Research/psextension/colorextension/lib/extension`
- Server returned status: 403, message: User is not allowed to modify using Asset/Project context

Have traced the signin process and can confirm correct user account and token was generated.

What are the permissions or services access I need to assign to the user to be able to change the context ?
Note: Tried using asset id and project id but both generate 403 error.

Thank you

**Comment by pmconne** [comment]
@paulius-valiunas?

**Reply by paulius-valiunas** [reply]
Sorry, this error message is a bit misleading. Extensions are now stored per team instead of per project. You can either supply your team id, or skip the `--contextId` argument and extension-cli will detect the default team of the logged in user.

**Reply by paulius-valiunas** [reply]
To be precise, they're stored/created per team, but can be accessed with a team/project/asset id.

**Reply by griemens** [reply]
Thank you for the clear information.
Have seen in the API somewhere how to query for a team id.

Note: From the documentation it was mentioned when Context Id is not provided, the extension will become publicly available. Which I thought would be all iTwinJS users and applications.
 
Will give it a test run and get back to share my findings.

Thank you
gert

**Reply by paulius-valiunas** [reply]
Could you post a link to the documentation page that says that? It might be for an old version. Currently in the readme it says:

> publish
> --extensionName (--en, -n) - [Optional] extension name. If not provided, will attempt to read from package.json in current working directory.
> --extensionVersion (--ev, -v) - [Optional] extension version. If not provided, will attempt to read from package.json in current working directory.
> --contextId (--cid) - [Optional] context Id to publish to. Should be a Team Id for private extensions. If not provided, will publish to the logged in user's default Team.
> --public (-p) - [Optional] set this flag to ignore contextId and attempt to publish a public extension.
> --filePath (--path) - path to a directory containing extension files to be uploaded.

https://github.com/imodeljs/imodeljs/tree/master/tools/extension-cli#command-options
So if you want to publish a public extension, you have to use the `--public` flag. Otherwise it will publish to your default team.

**Reply by calebmshafer** [reply]
To clarify a bit, the `--public` flag will currently not work as public Extensions are not yet supported by the Extension Service for anyone to contribute. This will be a future enhancement but I do not currently have a roadmap for when it will be available.

The "public" flag is for publishing an Extension so that anyone, who has an IMS/Bentley account, can access and use it. Whereas, without that flag it will publish, as @paulius-valiunas mentions, for you specific Team/Organization.
